Muni tech writers:  Ansible include_task search documentation incorrect #357 (#1263)
ansible/ansible-documentation: docs/docsite/rst/playbook_guide/playbook_pathing.rst

Task paths
==========

Relative paths used in a task typically refer to remote files and directories on the managed nodes. However, paths passed to lookup plugins and some paths used in action plugins such as the "src" path for the :ansplugin:``template <ansible.builtin.template#module>`` and :ansplugin:``copy <ansible.builtin.copy#module>`` modules refer to local files and directories on the control node.

How lookup and action plugins resolve local relative paths

> ----------------------------------------------------------

Reviewer: Reading the comment from @bcoca here: https://github.com/ansible/ansible-documentation/pull/1263#discussion_r1567576203

Seems like this heading is somewhat restrictive still. Maybe we could broaden it by changing to something like:

"How plugins resolve local relative paths"

When you specify a relative path for a local file, Ansible will try to find that file first in the current task's role, then in other roles that included or depend on the current role, then relative to the file in which the task is defined, and finally relative to the current play. It will take the first matching file that it finds. This way, if multiple files with the same file name exist, Ansible will find the file that is closest to the current task and that is most likely to be file you wanted.

Specifically, Ansible tries to find the file in the following order:

1. In the current role.

   1. In its appropriate subdirectory—"files", "vars", "templates" or "tasks", depending on the kind of file Ansible is searching for.
   2. Directly in its directory.
   
2. Like 1, in the parent role that called into this current role with ``include_role``, ``import_role``, or with a role dependency. If the parent role has its own parent role, Ansible will repeat this step with that role.
3. Like 1, in the current task file's directory.
4. Like 1, in the current play file's directory.

> Ansible does not search the current working directory. (The directory you're in when you execute Ansible.).

Reviewer: ```suggestion
Ansible does not search for local files in the current working directory, in other words the directory from which you execute Ansible.
```

Resolving local relative paths for task files

> ---------------------------------------------

Reviewer: Consider dropping this heading or nest it under the preceding one. TBH I think a separate heading is overkill and you could fit the details into a note.

> Ansible searches for task files within a role only if you include them with an ``include_role`` or ``import_role`` statement. If you use ``include``, ``include_task``, or ``import_task`` statement to include only the tasks from a specific file, but not the full role, Ansible searches for the task in the current role.

Reviewer: When resolving local relative paths for tasks files, Ansible uses the context of the role that includes tasks with an ``include_role`` or ``import_role`` statement.

Reviewer: Maybe something like that might be clearer for the first sentence and (again) match the comment here: https://github.com/ansible/ansible-documentation/pull/1263#discussion_r1567565138

As mentioned above this could fit into a note rather than a new heading.

In this case, Ansible tries to find the file in the following order:

1. In the current rule.

   1. In the ``tasks`` directory.
   2. Directly in its directory.

2. Like 1, in the current task file's directory.

> 3. Like 1, in the current play file's directory.

Reviewer: Could probably drop lines 47 to 55 here tbh.

Debugging search paths

> ----------------------

Reviewer: This should be a subheading / section heading, not at the same level as the preceding one.

Relative paths used in a task typically refer to remote files and directories on the managed nodes. However, paths passed to lookup plugins and some paths used in action plugins such as the "src" path for the :ansplugin:`template <ansible.builtin.template#module>` and :ansplugin:`copy <ansible.builtin.copy#module>` modules refer to local files and directories on the control node.

> Resolving local lookup and action plugin relative paths

Reviewer: ```suggestion
How lookup and action plugins resolve local relative paths
```

Reviewer 2: Maybe it's because English isn't my native language, but I find the 'for' a bit confusing, as if this is about the paths of the plugins, and not about how such plugins resolve local paths.

How about replacing 'for' by 'in', or rearranging the sentence to something like 'How lookup and action plugins resolve local relative paths'?

> Resolving local lookup and action plugin relative paths

Reviewer 3: this is not limited to those cases, any plugin that attempts a file access when called from within a task will have this issue, which adds filter and test plugins to the mix also (see `exists` test for an example), also connection and vars plugins.

This won't affect callbacks, inventory and many other plugins like that, which are more 'global' or execute independent of each task.

-------------------------------------------------------

When you specify a relative path for a local file, Ansible will try to find that file first in the current task's role, then in other roles that included or depend on the current role, then relative to the file in which the task is defined, and finally relative to the current play. It will take the first matching file that it finds. This way, if multiple files with the same file name exist, Ansible will find the file that is closest to the current task and that is most likely to be file you wanted.

> 

Reviewer: nit. don't think we need to add a line here.

3. Like 1, in the current task file's directory.
4. Like 1, in the current play file's directory.

Ansible does not search the current working directory. (The directory you're in when you execute Ansible.).

> Resolving local relative tasks paths

Reviewer: ```suggestion
Resolving local relative paths for task files
```

Reviewer 2: If you change this to 'paths for tasks', please better use 'paths for task files'. Otherwise it's a bit unclear whether this refers to task files or to paths mentioned in tasks (which are already covered by the action / lookup section above).

------------------------------------

> Ansible will only search for task files within a role if you actually included them with an `include_role` or `import_role`. If you instead use `include`, `include_task` or `import_task` to include just the tasks from a specific file but not the full role, Ansible will only search for the task in the current role.

Reviewer: ```suggestion
Ansible searches for task files within a role only if you include them with an `include_role` or `import_role` statement. If you use `include`, `include_task`, or `import_task` statement to include only the tasks from a specific file, but not the full role, Ansible searches for the task in the current role.
```

Reviewer: Some nits for grammar and clarity. In general it's better to use present simple tense instead of future.

Reviewer 2: Please also replace single backticks by double backticks, and please also in line 38 (it's not possible to add a comment there).

Reviewer 3: so, this has always been misleading as if the include/import_tasks is within a role itself, that role's context will be used. For example the RoleB's context will not be used if from a play or roleA you include/import_tasks from roleB. While the roleB context will not be used, in the 2nd case, roleA's context is still valid.

> In this case, Ansible tries to find the file

Reviewer: ```suggestion
In this case, Ansible tries to find the file in the following order:
```

Reviewer: Can we also apply this to the previous sentence so both lead-in sentences are consistent?

1. In the current rule.

>    1. In the "tasks" directory.

Reviewer: ```suggestion
   1. In the ``tasks`` directory.
```

   2. Directly in its directory.

2. Like 1, in the current task file's directory.

> 3. Like 1, in the current play file's directory.

Reviewer: I'm not sure about these last two and need to check.
